### PR TITLE
add reuters21578 dataset

### PR DESCRIPTION
**Pull request by mariamabarham**
This PR adds the `Reuters_21578` dataset https://kdd.ics.uci.edu/databases/reuters21578/reuters21578.html  
#353 

The datasets is a lit of `.sgm` files which are a bit different from xml file indeed `xml.etree` couldn't be used to read files. I consider them as text file (to avoid using external library)  and read line by line (maybe there is a better way to do, happy to get your opinion on it)

In the Readme file 3 ways to split the dataset are given.:

- The Modified Lewis ("ModLewis") Split: train, test and unused-set

- The Modified Apte ("ModApte") Split : train, test and unused-set

- The Modified Hayes ("ModHayes") Split: train and test

Here I consider the last one as the readme file highlight that this split provides the ability to compare results with those of the 2 first splits.

### CODE REVIEW

File: datasets/reuters21578/reuters21578.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+from __future__ import absolute_import, division, print_function
+
+import csv
+import os
+
+import nlp
+
+
+_CITATION = """\
```
Comment (lhoestq) [inline datasets/reuters21578/reuters21578.py:27]: Could you format the citation as usual ?

File: datasets/reuters21578/reuters21578.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+                        while "</BODY>" not in line:
+                            text += line
+                            line = f.readline()
+                        
+                        yield new_id, {
+                                "lewis_split": lewis_split,
+                                "cgis_split": cgis_split,
+                                "old_id": old_id,
+                                "new_id":  new_id,
+                                "topics":  topics,
```
Comment (lhoestq) [inline datasets/reuters21578/reuters21578.py:314]: I can see double spaces here. Could you re-run `make style` ?

**Comment by mariamabarham** [comment]
> Awesome !
> Good job on parsing these files :O
> 
> Do you think it would be hard to get the two other split configurations ?

It shouldn't be that hard, I think I can consider different config names for each split

File: datasets/reuters21578/reuters21578.py (hunk 1 above)
Comment (lhoestq) [reply datasets/reuters21578/reuters21578.py:27]: Thanks for the formatting 👍 

There are still a few things we could do. As the citation is rendered in the viewer and on the website we should pay attention that it's all beautiful x)
- make sure that the citations have the same indentation
- put the commas at the end of the lines
- place the = character in a consistent position (sometimes it's vertically aligned, sometimes not)
- remove empty spaces before the `}` (see the 6th citation for example)

**Comment by lhoestq** [comment]
> > Awesome !
> > Good job on parsing these files :O
> > Do you think it would be hard to get the two other split configurations ?
> 
> It shouldn't be that hard, I think I can consider different config names for each split

Yes that would be perfect

File: reuters/dataset_infos.json [hunk 3]
```diff
@@ -0,0 +1 @@
...Sequence"}, "people": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "orgs": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "exchanges": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "date": {"dtype": "string", "id": null, "_type": "Value"}, "title": {"dtype": "string", "id": null, "_type": "Value"}}, "supervised_keys": null, "builder_name": "reuters21578", "config_name": "default", "version": {"version_str": "0.0.0", "description": null, "nlp_version_to_prepare": null, "major": 0, "minor": 0, "patch": 0}, "splits": {"test": {"name": "test", "num_bytes": 930193, "num_examples": 720, "dataset_name": "reuters21578"}, "train": {"name": "train", "num_bytes": 18050316, "num_examples": 18323, "dataset_name": "reuters21578"}}, "download_checksums": {"https://kdd.ics.uci.edu/databases/reuters21578/reuters21578.tar.gz": {"num_bytes": 8150596, "checksum": "3bae43c9b14e387f76a61b6d82bf98a4fb5d3ef99ef7e7075ff2ccbcf59f9d30"}}, "download_size": 8150596, "dataset_size": 18980509, "size_in_bytes": 27131105}}
```
Comment (lhoestq) [inline reuters/dataset_infos.json:1]: I guess this file and `reuters/dummy/0.0.0/dummy_data.zip` are supposed to be removed ?

File: datasets/reuters21578/reuters21578.py [hunk 4]
```diff
@@ -248,79 +306,247 @@ def _generate_examples(self, filepath, split):
...
+                            line = f.readline()
+                        elif line.startswith("<TITLE>"):
+                            title = line[7:-9]
+                            line = f.readline()
+                        elif "<BODY>" in line:
+                            text = line.split('<BODY>')[1]
+                            line = f.readline()
+                            while "</BODY>" not in line:
+                                text += line
+                                line = f.readline()
```
Comment (lhoestq) [inline datasets/reuters21578/reuters21578.py:532]: Maybe this part can be factorized for the three configurations ?
I don't have a strong opinion on it though

File: datasets/reuters21578/reuters21578.py [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+@inproceedings{LEWIS91c ,
+author = {David D. Lewis},
+title = {Evaluating Text Categorization},
+booktitle = {Proceedings of Speech and Natural Language Workshop},
+year = {1991},
+month = {feb},
+organization = {Defense Advanced Research Projects Agency},
+publisher = {Morgan Kaufmann},
+pages = {312--318}
+
```
Comment (lhoestq) [inline datasets/reuters21578/reuters21578.py:81]:
```suggestion
```

File: datasets/reuters21578/reuters21578.py [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+                            line = f.readline()
+                            while "</BODY>" not in line:
+                                text += line
+                                line = f.readline()
+                            
+                            yield new_id, {
+                                    "lewis_split": lewis_split,
+                                    "cgis_split": cgis_split,
+                                    "old_id": old_id,
+                                    "new_id":  new_id,
```
Comment (lhoestq) [inline datasets/reuters21578/reuters21578.py:538]: double spaces are back ^^

**Comment by mariamabarham** [comment]
closing this PR and opening a new one to fix the circle CI problems

**Pull request closed without merging**